query/src/main/scala/filodb/query/exec/rangefn/RateFunctions.scala

```diff
@@ -46,16 +46,81 @@ object RateFunctions {
       val scaledDelta = delta * (extrapolateToInterval / sampledInterval)
       // for rate, we need rate as a per-second value
       val result = if (isRate) (scaledDelta / (endTimestamp - startTimestamp) * 1000) else scaledDelta
-      Some(result)
+      result
+    }
+  }
+
+  def derivFunction(window: Window): Double = {
+    if (window.size < 2) {
+      Double.NaN  // cannot calculate result without 2 samples
+    } else {
+      // We pass in an arbitrary timestamp that is near the values in use
+      // to avoid floating point accuracy issues
+      linearRegression(window, window(0).timestamp)
+    }
+  }
+
+  /**
+    * Logic is kept consistent with Prometheus' linearRegression function in order to get consistent results.
+    * We can look at optimizations (if any) later.
+    */
+  def linearRegression(window: Window, interceptTime: Long, isPredict: Boolean = false): Double = {
+    var n = 0.0
+    var sumX = 0.0
+    var sumY = 0.0
+    var sumXY = 0.0
+    var sumX2 = 0.0
+    for (i <- 0 until window.size) {
+      var sample = window(i)
+      var x = (sample.timestamp-interceptTime).toDouble / 1000.0
+      n += 1.0
+      sumY += sample.value
+      sumX += x
+      sumXY += x * sample.value
+      sumX2 += x * x
+    }
+    val covXY = sumXY - sumX*sumY/n
+    val varX = sumX2 - sumX*sumX/n
+    val slope = covXY / varX
+    val intercept = sumY/n - slope*sumX/n // keeping it, needed for predict_linear function = slope*duration + intercept
+    slope
+  }
+
+  def instantValue(startTimestamp: Long,
+                       endTimestamp: Long,
+                       window: Window,
+                       isRate: Boolean): Double = {
```

Comment:
I wonder about this API.   Wouldn't it be better for there to be a separate rate function?  Why try to mix up the rate function and embed it inside of this?